File: packages/less/src/less/source-map-builder.test.js
import { describe, it, expect, beforeEach, vi } from 'vitest';
import SourceMapBuilderFactory from './source-map-builder.js';

describe('SourceMapBuilder', () => {
    let mockSourceMapOutput;
    let mockEnvironment;
    let SourceMapBuilder;
    let mockRootNode;
    let mockImports;

    beforeEach(() => {
        // Mock SourceMapOutput class
        mockSourceMapOutput = {
            toCSS: vi.fn(() => 'body { color: red; }'),
            sourceMap: '{"version":3,"sources":["test.less"],"mappings":"AAAA"}',
            sourceMapURL: 'test.css.map',
            normalizeFilename: vi.fn((filename) => filename.replace(/\\/g, '/')),
            removeBasepath: vi.fn((path) => path.replace('/project/', ''))
        };

        function MockSourceMapOutputClass() {
            return mockSourceMapOutput;
        }

        // Mock environment
        mockEnvironment = {
            encodeBase64: vi.fn((str) => Buffer.from(str).toString('base64'))
        };

        // Mock root node
        mockRootNode = {};

        // Mock imports
        mockImports = {
            contents: { 'test.less': 'body { color: red; }' },
            contentsIgnoredChars: {}
        };

        SourceMapBuilder = SourceMapBuilderFactory(MockSourceMapOutputClass, mockEnvironment);
    });

    describe('constructor', () => {
        it('should initialize with options', () => {
            const options = {
                sourceMapFilename: 'output.css.map',
                sourceMapURL: 'custom.map'
            };

            const builder = new SourceMapBuilder(options);
            expect(builder.options).toBe(options);
        });

        it('should handle empty options', () => {
            const builder = new SourceMapBuilder({});
            expect(builder.options).toEqual({});
        });
    });

    describe('toCSS', () => {
        let builder;
        let options;

        beforeEach(() => {
            options = {
                sourceMapFilename: 'output.css.map',
                sourceMapURL: 'custom.map',
                sourceMapOutputFilename: 'output.css',
                sourceMapBasepath: '/project/',
                sourceMapRootpath: '/assets/',
                outputSourceFiles: true,
                sourceMapGenerator: {},
                sourceMapFileInline: false,
                disableSourcemapAnnotation: false
            };
            builder = new SourceMapBuilder(options);
        });

        it('should create SourceMapOutput with correct options', () => {
            const css = builder.toCSS(mockRootNode, {}, mockImports);

            expect(mockSourceMapOutput.toCSS).toHaveBeenCalledWith({});
            expect(css).toContain('body { color: red; }');
        });

        it('should set sourceMap and sourceMapURL from output', () => {
            builder.toCSS(mockRootNode, {}, mockImports);

            expect(builder.sourceMap).toBe(mockSourceMapOutput.sourceMap);
            expect(builder.sourceMapURL).toBe(mockSourceMapOutput.sourceMapURL);
        });

        it('should normalize sourceMapInputFilename when provided', () => {
            builder.options.sourceMapInputFilename = 'C:\\project\\input.less';
            
            builder.toCSS(mockRootNode, {}, mockImports);

            expect(mockSourceMapOutput.normalizeFilename).toHaveBeenCalledWith('C:\\project\\input.less');
            expect(builder.sourceMapInputFilename).toBe('C:/project/input.less');
        });

        it('should remove basepath from sourceMapURL when both are defined', () => {
            builder.options.sourceMapBasepath = '/project/';
            mockSourceMapOutput.sourceMapURL = '/project/styles/output.css.map';
            
            builder.toCSS(mockRootNode, {}, mockImports);

            expect(mockSourceMapOutput.removeBasepath).toHaveBeenCalledWith('/project/styles/output.css.map');
            expect(builder.sourceMapURL).toBe('styles/output.css.map');
        });

        it('should not remove basepath when sourceMapBasepath is undefined', () => {
            builder.options.sourceMapBasepath = undefined;
            mockSourceMapOutput.sourceMapURL = '/project/styles/output.css.map';
            
            builder.toCSS(mockRootNode, {}, mockImports);

            expect(mockSourceMapOutput.removeBasepath).not.toHaveBeenCalled();
            expect(builder.sourceMapURL).toBe('/project/styles/output.css.map');
        });

        it('should not remove basepath when sourceMapURL is undefined', () => {
            builder.options.sourceMapBasepath = '/project/';
            mockSourceMapOutput.sourceMapURL = undefined;
            
            builder.toCSS(mockRootNode, {}, mockImports);

            expect(mockSourceMapOutput.removeBasepath).not.toHaveBeenCalled();
        });

        it('should append CSS appendage to result', () => {
            mockSourceMapOutput.sourceMapURL = 'test.css.map';
            
            const css = builder.toCSS(mockRootNode, {}, mockImports);

            expect(css).toBe('body { color: red; }/*# sourceMappingURL=test.css.map */');
        });
    });

    describe('getCSSAppendage', () => {
        let builder;

        beforeEach(() => {
            builder = new SourceMapBuilder({});
            builder.sourceMapURL = 'test.css.map';
        });

        it('should return source map comment with URL', () => {
            const appendage = builder.getCSSAppendage();
            expect(appendage).toBe('/*# sourceMappingURL=test.css.map */');
        });

        it('should return inline source map when sourceMapFileInline is true', () => {
            builder.options = { sourceMapFileInline: true };
            builder.sourceMap = '{"version":3}';
            mockEnvironment.encodeBase64.mockReturnValue('eyJ2ZXJzaW9uIjozfQ==');
            
            const appendage = builder.getCSSAppendage();
            
            expect(mockEnvironment.encodeBase64).toHaveBeenCalledWith('{"version":3}');
            expect(appendage).toBe('/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozfQ== */');
        });

        it('should return empty string when sourceMapFileInline is true but no sourceMap', () => {
            builder.options = { sourceMapFileInline: true };
            builder.sourceMap = undefined;
            
            const appendage = builder.getCSSAppendage();
            expect(appendage).toBe('');
        });

        it('should return empty string when disableSourcemapAnnotation is true', () => {
            builder.options = { disableSourcemapAnnotation: true };
            
            const appendage = builder.getCSSAppendage();
            expect(appendage).toBe('');
        });

        it('should return empty string when no sourceMapURL', () => {
            builder.sourceMapURL = undefined;
            
            const appendage = builder.getCSSAppendage();
            expect(appendage).toBe('');
        });

        it('should return empty string when sourceMapURL is null', () => {
            builder.sourceMapURL = null;
            
            const appendage = builder.getCSSAppendage();
            expect(appendage).toBe('');
        });

        it('should return empty string when sourceMapURL is empty string', () => {
            builder.sourceMapURL = '';
            
            const appendage = builder.getCSSAppendage();
            expect(appendage).toBe('');
        });
    });

    describe('getExternalSourceMap', () => {
        it('should return the sourceMap', () => {
            const builder = new SourceMapBuilder({});
            builder.sourceMap = '{"version":3}';
            
            expect(builder.getExternalSourceMap()).toBe('{"version":3}');
        });

        it('should return undefined when no sourceMap', () => {
            const builder = new SourceMapBuilder({});
            
            expect(builder.getExternalSourceMap()).toBeUndefined();
        });
    });

    describe('setExternalSourceMap', () => {
        it('should set the sourceMap', () => {
            const builder = new SourceMapBuilder({});
            const sourceMap = '{"version":3}';
            
            builder.setExternalSourceMap(sourceMap);
            
            expect(builder.sourceMap).toBe(sourceMap);
        });
    });

    describe('isInline', () => {
        it('should return true when sourceMapFileInline is true', () => {
            const builder = new SourceMapBuilder({ sourceMapFileInline: true });
            expect(builder.isInline()).toBe(true);
        });

        it('should return false when sourceMapFileInline is false', () => {
            const builder = new SourceMapBuilder({ sourceMapFileInline: false });
            expect(builder.isInline()).toBe(false);
        });

        it('should return falsy when sourceMapFileInline is undefined', () => {
            const builder = new SourceMapBuilder({});
            expect(builder.isInline()).toBeFalsy();
        });
    });

    describe('getSourceMapURL', () => {
        it('should return the sourceMapURL', () => {
            const builder = new SourceMapBuilder({});
            builder.sourceMapURL = 'test.css.map';
            
            expect(builder.getSourceMapURL()).toBe('test.css.map');
        });

        it('should return undefined when no sourceMapURL', () => {
            const builder = new SourceMapBuilder({});
            
            expect(builder.getSourceMapURL()).toBeUndefined();
        });
    });

    describe('getOutputFilename', () => {
        it('should return sourceMapOutputFilename from options', () => {
            const builder = new SourceMapBuilder({ sourceMapOutputFilename: 'output.css' });
            
            expect(builder.getOutputFilename()).toBe('output.css');
        });

        it('should return undefined when no sourceMapOutputFilename', () => {
            const builder = new SourceMapBuilder({});
            
            expect(builder.getOutputFilename()).toBeUndefined();
        });
    });

    describe('getInputFilename', () => {
        it('should return sourceMapInputFilename', () => {
            const builder = new SourceMapBuilder({});
            builder.sourceMapInputFilename = 'input.less';
            
            expect(builder.getInputFilename()).toBe('input.less');
        });

        it('should return undefined when no sourceMapInputFilename', () => {
            const builder = new SourceMapBuilder({});
            
            expect(builder.getInputFilename()).toBeUndefined();
        });
    });

    describe('integration scenarios', () => {
        it('should handle complete workflow with all options', () => {
            const options = {
                sourceMapFilename: 'output.css.map',
                sourceMapURL: 'custom.map',
                sourceMapOutputFilename: 'output.css',
                sourceMapInputFilename: 'input.less',
                sourceMapBasepath: '/project/',
                sourceMapRootpath: '/assets/',
                outputSourceFiles: true,
                sourceMapFileInline: false,
                disableSourcemapAnnotation: false
            };

            const builder = new SourceMapBuilder(options);
            mockSourceMapOutput.sourceMapURL = '/project/assets/output.css.map';
            
            const css = builder.toCSS(mockRootNode, {}, mockImports);

            expect(css).toBe('body { color: red; }/*# sourceMappingURL=assets/output.css.map */');
            expect(builder.sourceMap).toBe(mockSourceMapOutput.sourceMap);
            expect(builder.sourceMapURL).toBe('assets/output.css.map');
            expect(builder.getInputFilename()).toBe('input.less');
        });

        it('should handle inline source map workflow', () => {
            const options = { sourceMapFileInline: true };
            const builder = new SourceMapBuilder(options);
            builder.sourceMap = '{"version":3}';
            mockEnvironment.encodeBase64.mockReturnValue('eyJ2ZXJzaW9uIjozfQ==');
            
            const css = builder.toCSS(mockRootNode, {}, mockImports);

            expect(css).toBe('body { color: red; }/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozfQ== */');
            expect(builder.isInline()).toBe(true);
        });

        it('should handle disabled source map annotation', () => {
            const options = { disableSourcemapAnnotation: true };
            const builder = new SourceMapBuilder(options);
            builder.sourceMapURL = 'test.css.map';
            
            const css = builder.toCSS(mockRootNode, {}, mockImports);

            expect(css).toBe('body { color: red; }');
        });
    });
});

File: packages/less/src/less/less_go/source_map_builder_test.go
package less_go

import (
	"encoding/base64"
	"testing"
)

type MockBuilderEnvironment struct {
	EncodeBase64Func func(str string) string
}

func (m *MockBuilderEnvironment) EncodeBase64(str string) string {
	if m.EncodeBase64Func != nil {
		return m.EncodeBase64Func(str)
	}
	return base64.StdEncoding.EncodeToString([]byte(str))
}

// MockBuilderNode that implements Node interface for testing
type MockBuilderNode struct {
	Node
	GenCSSFunc func(context map[string]any, output *SourceMapOutput)
}

func NewMockBuilderNode() *MockBuilderNode {
	return &MockBuilderNode{
		Node: Node{},
	}
}

func (m *MockBuilderNode) GenCSSSourceMap(context map[string]any, output *SourceMapOutput) {
	if m.GenCSSFunc != nil {
		m.GenCSSFunc(context, output)
	}
}

func TestSourceMapBuilder(t *testing.T) {
	mockEnvironment := &MockBuilderEnvironment{}
	mockImports := &Imports{
		Contents:             map[string]string{"test.less": "body { color: red; }"},
		ContentsIgnoredChars: map[string]int{},
	}

	t.Run("constructor", func(t *testing.T) {
		t.Run("should initialize with options", func(t *testing.T) {
			options := SourceMapBuilderOptions{
				SourceMapFilename: "output.css.map",
				SourceMapURL:      "custom.map",
			}

			builder := NewSourceMapBuilder(options)
			if builder.options.SourceMapFilename != "output.css.map" {
				t.Errorf("SourceMapFilename = %v, want %v", builder.options.SourceMapFilename, "output.css.map")
			}
			if builder.options.SourceMapURL != "custom.map" {
				t.Errorf("SourceMapURL = %v, want %v", builder.options.SourceMapURL, "custom.map")
			}
		})

		t.Run("should handle empty options", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			if builder.options.SourceMapFilename != "" {
				t.Errorf("Expected empty SourceMapFilename, got %v", builder.options.SourceMapFilename)
			}
		})
	})

	t.Run("ToCSS", func(t *testing.T) {
		options := SourceMapBuilderOptions{
			SourceMapFilename:              "output.css.map",
			SourceMapURL:                   "custom.map",
			SourceMapOutputFilename:        "output.css",
			SourceMapBasepath:              "/project/",
			SourceMapRootpath:              "/assets/",
			OutputSourceFiles:              true,
			SourceMapGenerator:             map[string]any{},
			SourceMapFileInline:            false,
			DisableSourcemapAnnotation:     false,
		}
		builder := NewSourceMapBuilder(options)

		t.Run("should create SourceMapOutput and return CSS", func(t *testing.T) {
			mockNode := &MockBuilderNode{
				Node: *NewNode(),
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			css := builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			if !contains(css, "body { color: red; }") {
				t.Errorf("CSS should contain generated content: %v", css)
			}
		})

		t.Run("should set sourceMap and sourceMapURL from output", func(t *testing.T) {
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			if builder.sourceMap == "" {
				t.Error("sourceMap should be set")
			}
		})

		t.Run("should normalize sourceMapInputFilename when provided", func(t *testing.T) {
			builder.options.SourceMapInputFilename = "C:\\project\\input.less"
			
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			expected := "/assets/C:/project/input.less"
			if builder.sourceMapInputFilename != expected {
				t.Errorf("sourceMapInputFilename = %v, want %v", builder.sourceMapInputFilename, expected)
			}
		})

		t.Run("should remove basepath from sourceMapURL when both are defined", func(t *testing.T) {
			options := SourceMapBuilderOptions{
				SourceMapBasepath: "/project/",
				SourceMapURL:      "/project/styles/output.css.map",
			}
			builder := NewSourceMapBuilder(options)
			
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			expected := "styles/output.css.map"
			if builder.sourceMapURL != expected {
				t.Errorf("sourceMapURL = %v, want %v", builder.sourceMapURL, expected)
			}
		})

		t.Run("should not remove basepath when sourceMapBasepath is empty", func(t *testing.T) {
			options := SourceMapBuilderOptions{
				SourceMapBasepath: "",
				SourceMapURL:      "/project/styles/output.css.map",
			}
			builder := NewSourceMapBuilder(options)
			
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			expected := "/project/styles/output.css.map"
			if builder.sourceMapURL != expected {
				t.Errorf("sourceMapURL = %v, want %v", builder.sourceMapURL, expected)
			}
		})

		t.Run("should append CSS appendage to result", func(t *testing.T) {
			options := SourceMapBuilderOptions{
				SourceMapURL: "test.css.map",
			}
			builder := NewSourceMapBuilder(options)
			
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			css := builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			expected := "body { color: red; }/*# sourceMappingURL=test.css.map */"
			if css != expected {
				t.Errorf("CSS = %v, want %v", css, expected)
			}
		})
	})

	t.Run("getCSSAppendage", func(t *testing.T) {
		builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
		builder.sourceMapURL = "test.css.map"

		t.Run("should return source map comment with URL", func(t *testing.T) {
			appendage := builder.getCSSAppendage(mockEnvironment)
			expected := "/*# sourceMappingURL=test.css.map */"
			if appendage != expected {
				t.Errorf("getCSSAppendage() = %v, want %v", appendage, expected)
			}
		})

		t.Run("should return inline source map when sourceMapFileInline is true", func(t *testing.T) {
			builder.options = SourceMapBuilderOptions{SourceMapFileInline: true}
			builder.sourceMap = `{"version":3}`
			mockEnvironment.EncodeBase64Func = func(str string) string {
				return "eyJ2ZXJzaW9uIjozfQ=="
			}
			
			appendage := builder.getCSSAppendage(mockEnvironment)
			
			expected := "/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozfQ== */"
			if appendage != expected {
				t.Errorf("getCSSAppendage() = %v, want %v", appendage, expected)
			}
		})

		t.Run("should return empty string when sourceMapFileInline is true but no sourceMap", func(t *testing.T) {
			builder.options = SourceMapBuilderOptions{SourceMapFileInline: true}
			builder.sourceMap = ""
			
			appendage := builder.getCSSAppendage(mockEnvironment)
			if appendage != "" {
				t.Errorf("getCSSAppendage() = %v, want empty string", appendage)
			}
		})

		t.Run("should return empty string when disableSourcemapAnnotation is true", func(t *testing.T) {
			builder.options = SourceMapBuilderOptions{DisableSourcemapAnnotation: true}
			
			appendage := builder.getCSSAppendage(mockEnvironment)
			if appendage != "" {
				t.Errorf("getCSSAppendage() = %v, want empty string", appendage)
			}
		})

		t.Run("should return empty string when no sourceMapURL", func(t *testing.T) {
			builder.sourceMapURL = ""
			
			appendage := builder.getCSSAppendage(mockEnvironment)
			if appendage != "" {
				t.Errorf("getCSSAppendage() = %v, want empty string", appendage)
			}
		})
	})

	t.Run("GetExternalSourceMap", func(t *testing.T) {
		t.Run("should return the sourceMap", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			builder.sourceMap = `{"version":3}`
			
			if got := builder.GetExternalSourceMap(); got != `{"version":3}` {
				t.Errorf("GetExternalSourceMap() = %v, want %v", got, `{"version":3}`)
			}
		})

		t.Run("should return empty string when no sourceMap", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			
			if got := builder.GetExternalSourceMap(); got != "" {
				t.Errorf("GetExternalSourceMap() = %v, want empty string", got)
			}
		})
	})

	t.Run("SetExternalSourceMap", func(t *testing.T) {
		t.Run("should set the sourceMap", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			sourceMap := `{"version":3}`
			
			builder.SetExternalSourceMap(sourceMap)
			
			if builder.sourceMap != sourceMap {
				t.Errorf("sourceMap = %v, want %v", builder.sourceMap, sourceMap)
			}
		})
	})

	t.Run("IsInline", func(t *testing.T) {
		t.Run("should return true when sourceMapFileInline is true", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{SourceMapFileInline: true})
			if !builder.IsInline() {
				t.Error("IsInline() should return true")
			}
		})

		t.Run("should return false when sourceMapFileInline is false", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{SourceMapFileInline: false})
			if builder.IsInline() {
				t.Error("IsInline() should return false")
			}
		})

		t.Run("should return false when sourceMapFileInline is not set", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			if builder.IsInline() {
				t.Error("IsInline() should return false when not set")
			}
		})
	})

	t.Run("GetSourceMapURL", func(t *testing.T) {
		t.Run("should return the sourceMapURL", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			builder.sourceMapURL = "test.css.map"
			
			if got := builder.GetSourceMapURL(); got != "test.css.map" {
				t.Errorf("GetSourceMapURL() = %v, want %v", got, "test.css.map")
			}
		})

		t.Run("should return empty string when no sourceMapURL", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			
			if got := builder.GetSourceMapURL(); got != "" {
				t.Errorf("GetSourceMapURL() = %v, want empty string", got)
			}
		})
	})

	t.Run("GetOutputFilename", func(t *testing.T) {
		t.Run("should return sourceMapOutputFilename from options", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{SourceMapOutputFilename: "output.css"})
			
			if got := builder.GetOutputFilename(); got != "output.css" {
				t.Errorf("GetOutputFilename() = %v, want %v", got, "output.css")
			}
		})

		t.Run("should return empty string when no sourceMapOutputFilename", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			
			if got := builder.GetOutputFilename(); got != "" {
				t.Errorf("GetOutputFilename() = %v, want empty string", got)
			}
		})
	})

	t.Run("GetInputFilename", func(t *testing.T) {
		t.Run("should return sourceMapInputFilename", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			builder.sourceMapInputFilename = "input.less"
			
			if got := builder.GetInputFilename(); got != "input.less" {
				t.Errorf("GetInputFilename() = %v, want %v", got, "input.less")
			}
		})

		t.Run("should return empty string when no sourceMapInputFilename", func(t *testing.T) {
			builder := NewSourceMapBuilder(SourceMapBuilderOptions{})
			
			if got := builder.GetInputFilename(); got != "" {
				t.Errorf("GetInputFilename() = %v, want empty string", got)
			}
		})
	})

	t.Run("integration scenarios", func(t *testing.T) {
		t.Run("should handle complete workflow with all options", func(t *testing.T) {
			options := SourceMapBuilderOptions{
				SourceMapFilename:              "output.css.map",
				SourceMapURL:                   "/project/assets/output.css.map",
				SourceMapOutputFilename:        "output.css",
				SourceMapInputFilename:         "input.less",
				SourceMapBasepath:              "/project/",
				SourceMapRootpath:              "/assets/",
				OutputSourceFiles:              true,
				SourceMapFileInline:            false,
				DisableSourcemapAnnotation:     false,
			}

			builder := NewSourceMapBuilder(options)
			
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			css := builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			expected := "body { color: red; }/*# sourceMappingURL=assets/output.css.map */"
			if css != expected {
				t.Errorf("CSS = %v, want %v", css, expected)
			}
			if builder.sourceMap == "" {
				t.Error("sourceMap should be set")
			}
			if builder.GetInputFilename() != "/assets/input.less" {
				t.Errorf("GetInputFilename() = %v, want %v", builder.GetInputFilename(), "/assets/input.less")
			}
		})

		t.Run("should handle inline source map workflow", func(t *testing.T) {
			options := SourceMapBuilderOptions{SourceMapFileInline: true}
			builder := NewSourceMapBuilder(options)
			builder.sourceMap = `{"version":3}`
			mockEnvironment.EncodeBase64Func = func(str string) string {
				return "eyJ2ZXJzaW9uIjozfQ=="
			}
			
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			css := builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			expected := "body { color: red; }/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozfQ== */"
			if css != expected {
				t.Errorf("CSS = %v, want %v", css, expected)
			}
			if !builder.IsInline() {
				t.Error("IsInline() should return true")
			}
		})

		t.Run("should handle disabled source map annotation", func(t *testing.T) {
			options := SourceMapBuilderOptions{DisableSourcemapAnnotation: true}
			builder := NewSourceMapBuilder(options)
			builder.sourceMapURL = "test.css.map"
			
			mockNode := &MockBuilderNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			css := builder.ToCSS(mockNode, map[string]any{}, mockImports, mockEnvironment)

			expected := "body { color: red; }"
			if css != expected {
				t.Errorf("CSS = %v, want %v", css, expected)
			}
		})
	})
}

// Helper function to check if a string contains a substring
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || s[:len(substr)] == substr || s[len(s)-len(substr):] == substr || containsSubstring(s, substr))
}

func containsSubstring(s, substr string) bool {
	if len(substr) > len(s) {
		return false
	}
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}